deepchem/hyper/gaussian_process.py

```diff
@@ -5,238 +5,249 @@
 import numpy as np
 import tempfile
 import os
-from deepchem.hyper.grid_search import HyperparamOpt
+import deepchem
+from deepchem.hyper.base_classes import HyperparamOpt
 from deepchem.utils.evaluate import Evaluator
-from deepchem.molnet.run_benchmark_models import benchmark_classification, benchmark_regression
 
 logger = logging.getLogger(__name__)
 
 
+def _convert_hyperparam_dict_to_filename(hyper_params):
+  """Helper function that converts a dictionary of hyperparameters to a string that can be a filename.
+
+  Parameters
+  ----------
+  hyper_params: dict
+    Maps string of hyperparameter name to int/float/list.
+
+  Returns
+  -------
+  filename: str
+    A filename of form "_key1_value1_value2_..._key2..."
+  """
+  filename = ""
+  keys = sorted(hyper_params.keys())
+  for key in keys:
+    filename += "_%s" % str(key)
+    value = hyper_params[key]
+    if isinstance(value, int):
+      filename += "_%s" % str(value)
+    else:
+      filename += "_%.2f" % value
+  return filename
+
+
+def compute_parameter_range(params_dict, search_range):
+  """Convenience Function to compute parameter search space.
+
+  Parameters
+  ----------
+  params_dict: dict
+    Dictionary mapping strings to Ints/Floats/Lists. For those
+    parameters in which int/float is specified, an explicit list of
+    parameters is computed with `search_range`.
+  search_range: int(float) (default 4)
+    For int/float values in `params_dict`, computes optimization range
+    on `[initial values / search_range, initial values *
+    search_range]`
+
+  Returns
+  -------
+  param_range: dict 
+    Dictionary mapping hyperparameter names to tuples. Each tuple is
+    of form `(value_type, value_range)` where `value_type` is a string
+    that is either "int" or "cont" and `value_range` is a list of two
+    elements of the form `[low, hi]`
+  """
+  # Range of optimization
+  param_range = {}
+  for hp, value in params_dict.items():
+    if isinstance(value, int):
+      value_range = [value // search_range, value * search_range]
+      param_range[hp] = ("int", value_range)
+      pass
+    elif isinstance(value, float):
+      value_range = [value / search_range, value * search_range]
+      param_range[hp] = ("cont", value_range)
+      pass
+    return param_range
+
+
 class GaussianProcessHyperparamOpt(HyperparamOpt):
   """
   Gaussian Process Global Optimization(GPGO)
+
+  This class uses Gaussian Process optimization to select
+  hyperparameters. Underneath the hood it uses pyGPGO to optimize
+  models. If you don't have pyGPGO installed, you won't be able to use
+  this class.
+
+  Note that `params_dict` has a different semantics than for
+  `GridHyperparamOpt`. `param_dict[hp]` must be an int/float and is
+  used as the center of a search range.

```

Comment:
The documentation in this file is inconsistent.  Sometimes it says int/float/list, and other times only int/float.  I think the latter is the correct one?  Explain this in more detail: GPGO can only optimize numerical hyperparameters, so all others have to be fixed.  And give an example showing how to specify both the fixed and variable hyperparameters.  For example,

 ```python
def model_builder(**model_params):
  n_layers = model_params['layers']
  layer_width = model_params['width']
  dropout = model_params['dropout']
  return dc.models.MultitaskClassifier(
    n_tasks=5,
    n_features=100,
    layer_sizes=[layer_width]*n_layers,
    dropouts=dropout
  )
```

On the other hand, I'm not sure this is the best design.  It would be simpler for users if they could specify arbitrary parameters of arbitrary type.  Perhaps `hyperparam_search` should let you specify two dicts: one for parameters to optimize, and one for parameters that are just passed through unchanged?  Or there could be an optional argument taking the names of the parameters to optimize (with the default being to optimize everythin)?  Or maybe `GaussianProcessHyperparamOpt` should permit you to include lists, but require they must have length 1 (only one possible value)?

Reply:
I did a bit of thinking and I think the current design with a custom `model_builder` is cleanest. This design matches what I've seen in other hyperparameter optimization libraries and allows for user flexibility in controlling arbitrary arguments to the model. I've added in the example you've suggested.

I've also fixed the documentation to only read int/float.